doc/conf.py

```diff
@@ -284,6 +284,7 @@ def get_min_setuptools_version():
 
 
 def config_defaults_from_declaration():
+    import ckan.cli.cli  # noqa: solves circular imports introduced by toolkit
```

Comment:
that's. wow. ok, really?

Reply:
YapXD 
When importing toolkit inside `ckan.plugins` we are breaking the rule "do not use the toolkit in CKAN core". For example, `ckan.logic` uses `ckan.plugins.IActions` interface. This means the toolkit is imported as a side-effect when `ckan.logic` is loaded. But toolkit uses `ckan.logic.get_action` and we have a circular dependency.


Import of `ckan.cli.cli` fixes it it the following way:

* load cli
* cli loads config
* config loads `ckan.plugins`
* `ckan.plugins` loads interfaces first(they are available at this moment) and then loads toolkit
* toolkit loads `ckan.logic`
* `ckan.logic` uses interfaces that are already available, so `ckan.plugins` is not actually touched and we don't have a circular dependency


As an alternative, we can recommend stop using `toolkit` as object from `ckan.plugins` and import it exlicitely instead. But it's a breaking change. We can in addition add a patch like bellow to `ckan.plugins`, that solve circular dependency. But this will be a hack, anyway.

```
def __getattr__(name):
    if name == 'toolkit':
        import ckan.plugins.toolkit as t
        return t
     raise AttributeError(<message>)
```

Reply:
It works when you run `ckan run` CLI because cli loads `ckan.plugins` first and issues is solved. But it just came to my mind that using `uWSGI` will likely have the same issues...

So, what do you think, which one is better:
* stop using `import ckan.plugins as p; p.toolkit` (maybe too extreme)
* Use `__getattr__` hack with an extra deprecation notice, suggesting to import toolkit directly
* solve the issue in a similar manner for uWSGI script - just organize imports, so that `ckan.plugins` loads before `ckan.logic`
* close this PR

Reply:
I feel like the `__getattr__` hack for for ckan.plugins might be the least disruptive (and is still a lot smaller than the whole ckan.plugins.toolkit using it)